### ISSUE TITLE
Reorganize Evaluation Scripts

### ISSUE THREAD

**Issue by alexlafleur** [ag-csw/LDStreamHMMLearn#35]
To reuse evaluation scripts for various purposes, we need to do some changes:
1. we want to use Variable_holder as an object (instead of a static class) where we set all parameters used within the evaluation scripts. 
2. This object is passed into the script
3. we want to pass in the error function that should be used within the evaluation
    * Signature: err: ndarray x SpectralMM --> number
    * transition matrice and model; output is the error value

**Comment by alexlafleur** [comment]
Okay, I am mostly done with this except for the num_trajectories evaluation. There, I get an exception again because num_estimations becomes <= 0.
Nevertheless, I adapt the evaluation scripts now, so that the new, generalized script is called. Then, we can remove obsolete and duplicate code scripts. We should take a look at the num_trajectories case later when we need it.

**Comment by greenTara** [comment]
Yes let's not spend time on that *problem now.*

**Comment by alexlafleur** [comment]
I adapted the script for the bayes error calculation which is based on eta and scalewindow.
The first plot shows the old scripts error results (8 runs):
![dependence_bayes_error_mm_delta 0](https://cloud.githubusercontent.com/assets/4387137/20712794/ef8c3d1c-b645-11e6-91db-04503e0a76e9.png)


The second shows the new adapted scripts error results (8 runs):
![dependence_bayes_error_mm_delta 0_new](https://cloud.githubusercontent.com/assets/4387137/20712798/f3d3f702-b645-11e6-98e4-b4916ac83c15.png)


Do you think the results can be assumed the same? Since we're simulating and sampling in each run it is possible that we get very different results. Do you  think the difference in colour is because the number of runs (=8) is too small or do you think there is something going wrong?

**Comment by greenTara** [comment]
We do want to have the possibility of setting the seed of the random number generator, and that would solve questions like this. But probably it would take another change of code to do that. How difficult do you think it would be to set that up?

**Comment by greenTara** [comment]
I do think it is most likely caused by the small number of runs. Another check would be to increase the number of runs and see if things get closer.

**Comment by alexlafleur** [comment]
I think it would not be a difficult change. Anyways, I would prefer to complete that issue so far before making another change. Many code fragments are dependent on that. Once we have cleaner code it is also easier to adapt that change. Maybe open a new issue for that, so we can keep that in mind?

**Comment by alexlafleur** [comment]
For numruns = 16 it came a little closer, I think.

![dependence_bayes_error_mm_delta 0_16](https://cloud.githubusercontent.com/assets/4387137/20713457/a5af28aa-b648-11e6-9660-a0aa088506f1.png)

![dependence_bayes_error_mm_delta 0_16_new](https://cloud.githubusercontent.com/assets/4387137/20713458/a5b13d48-b648-11e6-9ea2-39a332263840.png)

**Comment by greenTara** [comment]
yes, let's open a new issue for the random number seed setting

**Comment by greenTara** [comment]
Looks closer for 16 runs. That is sufficient confirmation for me.

**Comment by greenTara** [comment]
Are you also able to evaluate for the other error functions (e.g. in timescales)?

**Comment by alexlafleur** [comment]
I did not work on that yet, because I wanted to make sure that the existing evaluations we have are still working with the adapted function. But I will work on that afterwards!

**Comment by alexlafleur** [comment]
## QMM delta=0.5 numruns=1
The first plot is the (adapted) code we had to far are.
![dependence_bayes_error_qmm_delta 0 5](https://cloud.githubusercontent.com/assets/4387137/20798928/e19db086-b7e0-11e6-936a-fa950280c420.png)

The second is the one with the generalized evaluation script.
![dependence_bayes_error_qmm_delta_new 0 5](https://cloud.githubusercontent.com/assets/4387137/20798933/e471fd08-b7e0-11e6-8438-a1a223858639.png)

Question: There is this strange behaviour in the right upper corner in both diagrams. Do you have an idea where this is coming from?
  - I already checked that the parameters are all the same for each of the heatmap cells in both scripts.
  - I realized that the model transition matrix was NOT the same for the same k in both scripts.. Is that expected? (We are calling scaled_model.eval(k).trans)

**Comment by greenTara** [comment]
There were some problems with evaluating the transmission matrix at the right time value (#36). Has that been fixed? If not, then we cannot expect these scripts to give reasonable output.

**Comment by greenTara** [comment]
Also, is the seed for the random number generator being set? Since there is only run, we would expect different transition matrices in the two cases unless the seed is set.

**Comment by greenTara** [comment]
Another point to check - we should be using one simulation for the entire heatmap (when numruns = 1) - has that change been fully incorporated into the nonstationary case?

**Comment by greenTara** [comment]
The behavior in the last column is not necessarily wrong - with non-zero delta, we will start to see the effect of the non-stationary behavior, which will in some cases behave in the opposite manner than the stationary case. For example, in the stationary case, making window_size and shift bigger will always reduce the error, because more samples are included in the calculation. But in the nonstationary case,  these samples are being taken from different models, so increasing window_size and shift can make the error worse when there is a larger degree of drift within the window.

**Comment by greenTara** [comment]
What surprises me is that the third column behaves differently than the other two columns. I had expected that the parameters had been coordinated in a way that this would not happen. For this reason, I suspect that the issue #36 is coming into play.

**Comment by alexlafleur** [comment]
Oh, no I did not take a look at  #36 yet, so I'll do that now..

**Comment by alexlafleur** [comment]
Completed and testes transformations are:
* stationary Bayes Error Calculation for taumeta vs (eta, scalewindow)
* non-stationary Bayes Error Calculation for taumeta vs (eta, scalewindow)
* comparison of naive and bayes error calculation for taumeta vs (eta, scalewindow) for MM and QMM with delta=0

*Question:* Anything else that needs to be transformed NOW (that is urgent?)

Postponed transformations are:
* statconc evaluation
* timescaledisp evaluation

**Comment by greenTara** [comment]
As far as I can tell, this takes care of the heatmap requirements. So we should generate those plots and upload to sharelatex.

The other kind of plot is the one where time is on the horizontal axis (e.g. the deciles). We need to look at those scripts to see what needs to be done to finalize those.

**Comment by greenTara** [comment]
Here's a list of the taumeta heatmaps that we need:

MM only:
1. Taumeta heatmaps. Comparison of Naïve and Bayes error.
2. Taumeta heatmaps. Comparison of Naïve and Bayes performance.

MM and QMM
3. taumeta heatmaps: Comparison of error for MM and QMM with delta = 0. For testing only, not for paper.

QMM only
4.  taumeta heatmaps: Comparison of error for naïve and bayes, delta = 1

**Comment by greenTara** [comment]
The generation of individual plots is being carried on in separate issues, one for each plot.
